src/lib.rs

```diff
@@ -505,6 +526,49 @@ where
     }
 }
 
+impl<T, Q, N> Index<&'_ Q> for Counter<T, N>
+where
+    T: Hash + Eq + Borrow<Q>,
+    Q: Hash + Eq,
```

Comment:
I haven't previously played around much with the `Borrow` trait. I _think_ this reads that anything whose borrow is of type `Q` can be used as an index. This makes things generic over both `char` and `&char` in the example. Is this correct?

If yes, great, but I'd like to see an example in the tests. If no, please help me understand what the usage of `Borrow` does here.

Reply:
No. [`T: Borrow<T>` and `&'_T: Borrow<T>` for all `T` but not `T: Borrow<&'_ T>`](https://doc.rust-lang.org/stable/std/borrow/trait.Borrow.html#implementors).

```rust
impl<T, Q, N> Index<Q> for Counter<T, N>
where
    T: Hash + Eq + Borrow<Q>,
    Q: Hash + Eq,
    N: Zero,
{
    type Output = N;

    /// Index in immutable contexts
    ///
    /// ```
    /// # use counter::Counter;
    /// let counter = Counter::<_>::init("aabbcc".chars());
    /// assert_eq!(counter[&'a'], 2);
    /// assert_eq!(counter[&'b'], 2);
    /// assert_eq!(counter[&'c'], 2);
    /// assert_eq!(counter[&'d'], 0);
    /// ```
    fn index(&self, key: Q) -> &N {
        self.map.get(&key).unwrap_or(&self.zero)
    }
}
```

```
error[E0308]: mismatched types
 --> src/lib.rs:562:9
  |
6 | counter[&'c'] += 1;
  |         ^^^^ expected char, found &char
  |
  = note: expected type `char`
             found type `&char`
```

<https://doc.rust-lang.org/stable/std/borrow/trait.Borrow.html#examples>

Reply:
It is possible to accept both of `char` and `&chars` like [`defaultmap::DefaultHashMap::get`](https://docs.rs/defaultmap/0.4.0/defaultmap/struct.DefaultHashMap.html#method.get).

Reply:
It was not possible.

```
error[E0207]: the type parameter `Q` is not constrained by the impl trait, self type, or predicates
   --> src/lib.rs:539:9
    |
539 | impl<T, Q, R, N> Index<R> for Counter<T, N>
    |         ^ unconstrained type parameter

error: aborting due to previous error

For more information about this error, try `rustc --explain E0207`.
```